Composer/packages/server/src/models/bot/builder.ts

```diff
@@ -101,6 +109,73 @@ export class Builder {
     this._locale = v;
   }
 
+  public getModelPathAsync = async () => {
+    let appDataPath = '';
+    if (process?.versions?.hasOwnProperty('electron')) {
+      const { app } = await import('electron');
+      appDataPath = app.getPath('appData');
+    } else {
+      appDataPath = process.env.APPDATA || process.env.HOME || '';
+    }
+    const baseModelPath = Path.resolve(appDataPath, 'BotFrameworkComposer', 'models');
+    return baseModelPath;
+  };
+
+  /**
+   * Orchestrator: Perform the full build process
+   * 1) Query the Orchestrator service for the latest default model
+   * 2) If it has changed or never been downloaded, download it to user's AppData/BotFrameworkComposer folder
+   * 3) Generate the embedding/snapshot data for Orchestrator (.blu files) and place in /generated folder
+   * 4) Generate settings file for runtime to lookup model and snapshot paths
+   * @param luFiles LU Files needed to build snapshot data
+   */
+  public runOrchestratorBuild = async (luFiles: FileInfo[]) => {
+    if (!luFiles.length) return;
+
+    const nlrList = await this.runOrchestratorNlrList();
+    const defaultNLR = nlrList.default;
+    const modelPath = Path.resolve(await this.getModelPathAsync(), defaultNLR.replace('.onnx', ''));
+
+    if (!existsSync(modelPath)) {
+      const handler: IOrchestratorProgress = (status) => {
+        log(status);
+      };
+      await this.runOrchestratorNlrGet(modelPath, defaultNLR, handler, handler);
+    }
+
+    // build snapshots from LU files
+    const returnData = await this.orchestratorBuilder(luFiles, modelPath);
+
+    // write snapshot data into /generated folder
+    const snapshots: any = {};
+    for (const dialog of returnData.outputs) {
+      const bluFilePath = Path.resolve(this.generatedFolderPath, dialog.id.replace('.lu', '.blu'));
+      snapshots[dialog.id.replace('.lu', '').replace(/[-.]/g, '_')] = bluFilePath;
+
+      writeFile(bluFilePath, Buffer.from(dialog.snapshot), (err) => {
```

Comment:
why not using the "promisified" writeFile from fs-extra and await?

Reply:
Thanks @hatpick, I completely missed this.  Fixed this and the `pathExists` function on line 138.